Implement readCursor

mheinzel opened the pull request:
# Description

Depends on #381. The code isn't pretty yet, but should work.

jorisdral commented on src/Database/LSMTree/Internal/Vector.hs:
```diff
@@ -95,3 +96,24 @@ unsafeInsertWithMStrict ::
+      | n >= len = (, b) <$> V.unsafeFreeze vec
+      | otherwise =
+          f b >>= \case
+            (Nothing, b') ->
```
```suggestion
            (Nothing, !b') ->
```

jorisdral commented on src/Database/LSMTree/Internal/Vector.hs:
```diff
@@ -95,3 +96,24 @@ unsafeInsertWithMStrict ::
+            (Nothing, b') ->
+              (, b') <$> V.unsafeFreeze (VM.slice 0 n vec)
+            (Just a,  b') -> do
+              VM.unsafeWrite vec n a
```
Also, you might make this element strict, i.e. use `unsafeWrite vec n $! a`. This could also be reflected in the name -- and it might deserve a comment on just how strict the result actually is

mheinzel replied:
I did it by adding a bang on the pattern match on `a` and added a comment.

jorisdral commented on src/Database/LSMTree/Internal/Vector.hs:
```diff
@@ -95,3 +96,24 @@ unsafeInsertWithMStrict ::
+    go vec 0 b0
+  where
+    go vec !n !b
+      | n >= len = (, b) <$> V.unsafeFreeze vec
```
```suggestion
      | n >= len = (, b) <$!> V.unsafeFreeze vec
```

jorisdral commented on src/Database/LSMTree/Internal/Vector.hs:
```diff
@@ -95,3 +96,24 @@ unsafeInsertWithMStrict ::
+      | otherwise =
+          f b >>= \case
+            (Nothing, b') ->
+              (, b') <$> V.unsafeFreeze (VM.slice 0 n vec)
```
```suggestion
              (, b') <$!> V.unsafeFreeze (VM.slice 0 n vec)
```

jorisdral commented on src/Database/LSMTree/Monoidal.hs:
```diff
@@ -306,7 +307,10 @@ readCursor ::
-readCursor = undefined
+readCursor n (Internal.MonoidalCursor c) =
+    Internal.readCursor const n c $ \k v mblob ->
+      assert (null mblob) $
```
This should probably be an `error` instead of an `assert`, like we do in `toMonoidalLookupResult`

mheinzel replied:
Good catch!

mheinzel replied:
Not sure. It's an extra check for something that should never happen and we can handle it pretty gracefully. Similarly, we don't fail if there are mupserts when reading from a normal cursor, since the internal code just resolves it with the next entries, so it becomes indistinguishable from a single insert before being passed to the normal-specific code.

jorisdral commented on src/Database/LSMTree/Internal.hs:
```diff
@@ -869,6 +877,112 @@ closeCursor Cursor {..} = do
         V.forM_ cursorRuns $ freeTemp reg . Run.removeReference
         return CursorClosed
 
+readCursor ::
+     forall m h res. m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
```
Needs a `SPECIALISE`

jorisdral commented on src/Database/LSMTree/Monoidal.hs:
```diff
@@ -306,7 +307,10 @@ readCursor ::
   -> m (V.Vector (QueryResult k v))
-readCursor = undefined
+readCursor n (Internal.MonoidalCursor c) =
+    Internal.readCursor const n c $ \k v mblob ->
```
This should probably not be `const`

mheinzel replied:
Good catch!

jorisdral commented on test/Test/Database/LSMTree/Internal.hs:
```diff
@@ -207,13 +206,45 @@ prop_roundtripCursor kops = ioProperty $
+
+    expected =
+      V.fromList . mapMaybe (traverse entryToValue) $
+          Map.assocs . Map.fromListWith (combine appendSerialisedValue) $
```
Will this test ever encounter duplicates? It might be worth labelling so that we can see

mheinzel replied:
I added labelling, or rather tabulation to also see how many keys in a test run are duplicates and how often they occur.

```
        +++ OK, passed 100 tests.
        
        duplicates (46 in total):
        85% 2
        13% 3
         2% 7
```

So not often, but they exist.

jorisdral commented on src/Database/LSMTree/Internal.hs:
```diff
@@ -743,6 +748,8 @@ retrieveBlobs sesh wrefs =
   Cursors
 -------------------------------------------------------------------------------}
 
+-- TODO: Move to a separate Cursors module
```
:heavy_plus_sign:

mheinzel replied:
I will do it in a follow-up PR.

jorisdral reviewed (commented):
Not a full review yet, but here are some comments already

Let's also enable the statemachine tests

```diff
diff --git a/test/Test/Database/LSMTree/Normal/StateMachine.hs b/test/Test/Database/LSMTree/Normal/StateMachine.hs
index 0d3f4719..16311412 100644
--- a/test/Test/Database/LSMTree/Normal/StateMachine.hs
+++ b/test/Test/Database/LSMTree/Normal/StateMachine.hs
@@ -193,6 +193,7 @@ propLockstepIO_RealImpl_RealFS = testProperty "propLockstepIO_RealImpl_RealFS" $
       where
         handler' :: LSMTreeError -> Maybe Model.Err
         handler' ErrTableClosed = Just Model.ErrTableHandleClosed
+        handler' ErrCursorClosed = Just Model.ErrCursorClosed
         handler' (ErrSnapshotNotExists _snap) = Just Model.ErrSnapshotDoesNotExist
         handler' (ErrSnapshotExists _snap) = Just Model.ErrSnapshotExists
         handler' (ErrSnapshotWrongType _snap) = Just Model.ErrSnapshotWrongType
@@ -1033,8 +1034,7 @@ arbitraryActionWithVars _ findVars _st = QC.frequency $ concat [
       -> [(Int, Gen (Any (LockstepAction (ModelState h))))]
     withVars' genVar = [
           (2, fmap Some $ CloseCursor <$> (fromRight <$> genVar))
-          -- TODO: enable generators as we implement the actions for the /real/ lsm-tree
-        -- , (10, fmap Some $ ReadCursor <$> (QC.getNonNegative <$> QC.arbitrary) <*> (fromRight <$> genVar))
+        , (10, fmap Some $ ReadCursor <$> (QC.getNonNegative <$> QC.arbitrary) <*> (fromRight <$> genVar))
         ]

     withVars'' ::
```

It's probably also worth adding a few class-based tests for cursor, which can run against both the reference implementation and the real implementation

mheinzel commented:
I enabled the state machine tests and added class-based property tests. I also rebased and addressed some of the the feedback. The remainder will follow soon.

mheinzel commented:
I think this is ready, apart from probably rebasing onto #346 later.

I also ran the benchmarks from #377 

```
benchmarking Bench.Database.LSMTree.Normal/cursor-scan-vs-range-lookup-scan/cursor-scan
time                 10.32 ms   (10.14 ms .. 10.52 ms)
                     0.996 R²   (0.992 R² .. 0.999 R²)
mean                 10.50 ms   (10.39 ms .. 10.71 ms)
std dev              396.5 μs   (256.9 μs .. 609.4 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.903e7    (2.903e7 .. 2.903e7)
  y                  2620.153   (1709.444 .. 3490.414)
variance introduced by outliers: 16% (moderately inflated)
```

Which works out to reading about 7.75 million entries per second and 362 bytes/entry of allocations (keys and values are 8 byte each). Does that sound reasonable? Using `nfIO` instead of `whnfIO` doesn't change the result, so it seems like we are really forcing all the results.

dcoutts commented on test/Test/Database/LSMTree/Internal.hs:
```diff
@@ -207,13 +206,52 @@ prop_roundtripCursor kops = ioProperty $
+    chunkSize = 50
+    toResult k v b = (coerce k, (v, fmap toBlob b))
+
+    go !acc = do
+      res <- readCursor resolve chunkSize cursor toResult
+      if V.length res < chunkSize then return (acc <> res)
+                                  else go (acc <> res)
```
Quadratic append, but probably doesn't matter for test case sizes. Otherwise accumulate chunks and reverse & concat at the end.

mheinzel replied:
Yes, I just did the simplest thing here.

dcoutts commented on test/Test/Database/LSMTree/Normal/StateMachine.hs:
```diff
@@ -1033,8 +1034,7 @@ arbitraryActionWithVars _ findVars _st = QC.frequency $ concat [
           (2, fmap Some $ CloseCursor <$> (fromRight <$> genVar))
-          -- TODO: enable generators as we implement the actions for the /real/ lsm-tree
-        -- , (10, fmap Some $ ReadCursor <$> (QC.getNonNegative <$> QC.arbitrary) <*> (fromRight <$> genVar))
+        , (10, fmap Some $ ReadCursor <$> (QC.getNonNegative <$> QC.arbitrary) <*> (fromRight <$> genVar))
```
:tada:

dcoutts reviewed (approved):
Looks good!

mheinzel commented on src/Database/LSMTree/Internal.hs:
```diff
@@ -869,6 +877,115 @@ closeCursor Cursor {..} = do
+    -- * any function that doesn't take a 'hasMore' argument assumes that the
+    --   readers have not been drained yet, so we must check before calling them
+    -- * there is probably opportunity for optimisations
+    readEntries hfs hbio readers =
```
Makes sense, I'll take your patch.

dcoutts reviewed (commented):
LGTM!

jorisdral commented:
> ...
> Which works out to reading about 7.75 million entries per second and 362 bytes/entry of allocations (keys and values are 8 byte each). Does that sound reasonable? Using `nfIO` instead of `whnfIO` doesn't change the result, so it seems like we are really forcing all the results.

For these specific benchmarks, there is no difference because the benchmarked action returns `IO ()`

mheinzel commented:
> For these specific benchmarks, there is no difference because the benchmarked action has type `IO ()`

That's true, but when I tried this I changed `forM_` to `forM` as well.
